### PR TITLE
Reduce use of `AbsolutePath` with precondition failure

### PR DESCRIPTION
**Pull request by neonichu**
Since it comes up regularly as a perceived user facing crash, we should reduce our use of the plain `AbsolutePath` initializer. I actually don't think it makes sense to even provide a non-throwing initializer that takes anything but a `StaticString` and even that's debatable.

This doesn't resolve all possible uses and specifically excludes uses in tests, but does fix several user facing issues, e.g.

```
❯ export SWIFTPM_TESTS_MODULECACHE=blah
❯ swift test
zsh: illegal hardware instruction  swift test
```

### CODE REVIEW

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
Moved to draft since I'll be addressing the remaining call sites as well as discussed.

**Comment by neonichu** [comment]
https://github.com/apple/swift-tools-support-core/pull/353
@swift-ci please smoke test

**Comment by neonichu** [comment]
https://github.com/apple/swift-tools-support-core/pull/353
@swift-ci please smoke test

File: Sources/Basics/FileSystem+Extensions.swift [hunk 1]
```diff
@@ -21,11 +21,15 @@ import TSCBasic
...
     public var dotSwiftPM: AbsolutePath {
-        self.homeDirectory.appending(component: ".swiftpm")
+        get throws {
+            return try self.homeDirectory.appending(component: ".swiftpm")
+        }
     }
 
     fileprivate var idiomaticSwiftPMDirectory: AbsolutePath? {
-        return FileManager.default.urls(for: .libraryDirectory, in: .userDomainMask).first.flatMap { AbsolutePath($0.path) }?.appending(component: "org.swift.swiftpm")
+        get throws {
```
Comment (tomerd) [inline Sources/Basics/FileSystem+Extensions.swift:30]: nice!

**Review by tomerd — APPROVED** [review]

File: Tests/CommandsTests/PackageRegistryToolTests.swift [hunk 2]
```diff
@@ -52,7 +52,7 @@ final class PackageRegistryToolTests: CommandsTestCase {
     func testLocalConfiguration() throws {
         try fixture(name: "DependencyResolution/External/Simple") { fixturePath in
             let packageRoot = fixturePath.appending(component: "Bar")
-            let configurationFilePath = AbsolutePath(".swiftpm/configuration/registries.json", relativeTo: packageRoot)
+            let configurationFilePath = AbsolutePath(path: ".swiftpm/configuration/registries.json", relativeTo: packageRoot)
```
Comment (tomerd) [inline Tests/CommandsTests/PackageRegistryToolTests.swift:55]: is this throwing, or should be?

**Comment by neonichu** [comment]
https://github.com/apple/swift-driver/pull/1204
https://github.com/apple/swift-tools-support-core/pull/353
https://github.com/apple/sourcekit-lsp/pull/655
@swift-ci please smoke test

**Comment by neonichu** [comment]
https://github.com/apple/swift-driver/pull/1204
https://github.com/apple/swift-tools-support-core/pull/353
https://github.com/apple/sourcekit-lsp/pull/655
@swift-ci please smoke test

**Pull request merged**